# ACNE
Absolute Copy Number Estimation using low-coverage whole genome sequencing data

##### ACNE has arrived! ###### 
There are currently eight functions: acne, ploidyplotloop, ObjectsampleToTemplate, singlemodel, singleplot, getadjustedsegments, linkmutationdata, postanalysisloop, analyzegenomiclocations.# I have also added the "multiplot" function for making summary sheets.
acne takes a folder with rds-files (default, make sure all are segmented) or bam-files and returns plots for the most likely errors in convenient subfolders# in case of bam-files it will run binsizes 30, 100, 500 and 1000 kbp as default, but a vector of desired binsizes can be used as input# binsizes available are 1, 5, 10, 15, 30, 50, 100, 500, and 1000 kbp# note that output will be larger and programs will run longer with the smaller binsizes (1 kbp is pretty hardcore)# ploidies specifies for which ploidies fits will be made: default is 2 but you can input a vector with all desired ploidies, e.g. c(2,4)# beggers be choosers! acne likes pdf-files, but for large datasets or small binsizes you might want to go with 'png'# a standard method for error calculation is the root mean squared error (RMSE), but you can argue you should actually punish more the #   long segments that are a little bit off compared to the short segments that are way off# 'MAE' weighs every error equally, whereas 'SMRE' does the latter; these will generally generate more fits (I see that as a downside)#   but it is more likely that the segments of which you are sure are sticking tighter to the integer ploidy in the best fit# the parameter "penalty" sets a penalty for the error calculated at lower cellularity#   errors are divided by the cellularity to the power of the penalty: default = 0 (no penalty)# new functionality for large data sets (many samples!):# fits will always have to be chosen manually, but you can now open the file with all errorlists to pick the most likely fit without looking at the profile# use the generated tab-delimited file to fill in the column with the likely fit# added argument trncname (truncate name) which truncates the name to everything before the first instance of _ - set TRUE if applicable, #   or specify regular expression, e.g. "-.*" to REMOVE everything after and including the first dash found in a sample name# added argument printsummaries: superbig summary files may crash the program, so you can set this argument to FALSE or 2 if you still want the error plots
ploidyplotloop is the meat of acne, and it can be run as a separate function as well; # it takes a QDNAseq-object as input and also needs a folder to write the files to# this is particularly handy if you want to analyze a whole QDNAseq-object that is loaded in your R environment
ObjectsampleToTemplate is pretty much there to parse QDNAseqobjects into the dataframe structure used by the singlemodel and singleplot functions# these latter functions call ObjectsampleToTemplate itself when necessary, but it can be handy to make a template if you expect some repeated use of the functions# or if you want to make the template and then do your own obscure manipulations to it (I know you want to!)
singlemodel: like the name implies it runs the fitting algorithm on a single sample and spits out the info you want!# not strictly necessary to save it to a variable, but that might still be handy# singlemodel comes with the awesomeness of manual input: you can restrain the model to the ploidy you expect (default 2, but hey, ploidy 5 happens, right?)# and you can tell the program if you think there is a different standard (this should only be necessary when the median segment happens to be a subclonal variant; very rare)
singleplot: again, you already know what it does! But did you know you can change the chart title to anything you like? Well, I'm here to tell you it can# and there is more! Don't believe the cellularities the program calculated for you? You don't have to! Just fill in the cellularity you believe to be true, and singleplot will take your word for it!
getadjustedsegments: get info of the actual segments in a handy dataframe! Contains start and end location, "number of probes" (number of bins supporting the segment value),# value of the segment (Segment_Mean is the value from QDNAseq, Segment_Mean2 is calculated from adjustedcopynumbers within the segment),# and the nearest ploidy with the chance (log10 p-value) that the segment has this ploidy. A very low p-value indicates a high chance of subclonality, # although this value should be approached with extreme caution (and is at the moment not corrected for multiple testing)
linkmutationdata: now we're talking! Give a tab-delimited file with mutation data, and this function will tell you what the copy number is# at that genomic location, and it will guess how many copies are mutant! Read more about the options for this function in the code
postanalysisloop: you've run acne, you've picked your models, you have your mutation data ... Now if you could only ...# say no more! this function combines the power of all above functions and your own brain (you still choose the models)# only for the brave of heart
